### ISSUE TITLE
MVCC and `space:on_replace` dirty reads

### ISSUE THREAD

**Issue by Gerold103** [tarantool/tarantool#6246]
There are 2 issues. First is `space:insert()` somewhy hangs on conflict until the other transaction is committed. Secondly, I can leak dirty data via `space:on_replace()` triggers.

---
Example 1 about insertion hang:
```Lua
log = require('log')
fiber = require('fiber')

box.cfg({
    memtx_use_mvcc_engine = true,
})
s = box.schema.create_space('test')
_ = s:create_index('pk')
s:on_replace(function(old, new)
    log.info('%s %s', old, new)
end)


function do_replace_f(value)
    fiber.self():set_joinable(true)
    box.begin()
    log.info('after begin %s', value)
    s:insert{1, value}
    log.info('after insert %s', value)
    box.commit()
    log.info('after commit %s', value)
end

box.error.injection.set('ERRINJ_WAL_DELAY', true)
f1 = fiber.create(do_replace_f, 1)
f2 = fiber.create(do_replace_f, 2)

-- box.error.injection.set('ERRINJ_WAL_DELAY', false)
-- f1:join()
-- f2:join()
```
In this example the second fiber hangs on `s:insert{1, value}` until I finish the first transaction.
```
tarantool> f1 = fiber.create(do_replace_f, 1)
2021-07-25 18:38:22.012 [51165] main/118/lua I> after begin 1
2021-07-25 18:38:22.012 [51165] main/118/lua I> nil [1, 1]
2021-07-25 18:38:22.012 [51165] main/118/lua I> after insert 1
---
...

tarantool> f2 = fiber.create(do_replace_f, 2)
2021-07-25 18:38:22.625 [51165] main/119/lua I> after begin 2
---
...

```

---
Example 2 about dirty reads:
```Lua
log = require('log')
fiber = require('fiber')

box.cfg({
    memtx_use_mvcc_engine = true,
})
s = box.schema.create_space('test')
_ = s:create_index('pk')
s:on_replace(function(old, new)
    log.info('%s %s', old, new)
end)


function do_replace_f(value)
    fiber.self():set_joinable(true)
    s:replace{1, value}
end

box.error.injection.set('ERRINJ_WAL_DELAY', true)
f1 = fiber.create(do_replace_f, 1)
f2 = fiber.create(do_replace_f, 2)

-- box.error.injection.set('ERRINJ_WAL_DELAY', false)
-- f1:join()
-- f2:join()
```
Here the second fiber sees changes of the first fiber in `on_replace` function:
```
tarantool> f1 = fiber.create(do_replace_f, 1)
2021-07-25 18:41:52.431 [51231] main/118/lua I> nil [1, 1]
---
...

tarantool> f2 = fiber.create(do_replace_f, 2)
2021-07-25 18:41:53.049 [51231] main/119/lua I> [1, 1] [1, 2]
---
...
```
Look at the last log line - the tuple from another transaction is visible.

**Comment by Korablev77** [comment]
@Gerold103 
I guess this issue shouldn't be considered to be a bug (at least now). MVCC in these cases demonstrates read committed isolation level and it is expected behavior (but currently there are debates concerning it). I believe @alyapunov can provide mode details about the subj (or simply read the proposal document at hackmd).

**Comment by Gerold103** [comment]
But its commit wasn't done yet, was it? Commit is performed after WAL write. Look at another example, without `on_replace` triggers:
```Lua
log = require('log')
fiber = require('fiber')

box.cfg({
    memtx_use_mvcc_engine = true,
})
s = box.schema.create_space('test')
_ = s:create_index('pk')

function do_replace_f(value)
    fiber.self():set_joinable(true)
    s:replace{1, value}
end

box.error.injection.set('ERRINJ_WAL_DELAY', true)
f = fiber.create(do_replace_f, 1)

box.begin()
s:get{1}
box.commit()

s:get{1}
```
```
tarantool> box.begin()
---
...

tarantool> s:get{1}
---
- [1, 1]
...

tarantool> box.commit()
---
...

tarantool> s:get{1}
---
...
```
I see the tuple in `box.begin()`/`box.commit()` pair, but don't see when do a plain `get`. In both cases I got the dirty data and could already return it to a user.

**Comment by Korablev77** [comment]
Ooops, I meant "read-prepared", sorry for confusion:(

**Comment by alyapunov** [comment]
1. Example 1: I wrapped insert with pcall and printed result. The second fiber received error:
`false, Duplicate key exists in unique index "pk" in space "test" with old tuple - [1, 1] and new tuple - [1, 2]`. So the fiber did not hand but terminated after exception. I think the behavior is correct.

**Comment by Gerold103** [comment]
Well, that was the point. The second fiber was not supposed to see the old tuple until the first fiber's commit ends. But I suppose it is considered fine if you are using prepare-isolation, like Nikita said. So the changes are visible even before commit for reads at least. That is still a dirty read though, I don't see how can it be used safely if data can disappear after reading. Does not seem better than no mvcc at all.

**Comment by alyapunov** [comment]
Yes, that a kind of dirty read, but note that it appears only in case of error ('tuple exists' in that case). In the other 'normal' case that is not a dirty read: to the moment when (or if) the transaction is successfully completed that read is guarantee to be confirmed.
I've create #6930 that completely describes desired behavior and added a special paragraph for this case.

**Comment by alyapunov** [comment]
Example 2: the same idea: on_replace trigger is launched within a transaction that is definitely read-write and thus to the moment of commit this read will become confirmed. Even more, if that we try to read older, confirmed tuple - that will break serialization and the transaction will be aborted. Note that all reads of a transaction are not legitimate until the transaction is committed.

**Comment by alyapunov** [comment]
Example 3: (the difference between autocommit and explicit transaction select) - thanks, will be fixed in #6930.
